### PR TITLE
Ci: Scheduled Trivy Scans of Eraser Images

### LINKED ISSUE #783: track trivy eraser image vulnerabilities as a recurring job
**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Track trivy eraser image vulnerabilities more frequently.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Eraser version:
- Kubernetes version: (use `kubectl version`):

### PR DESCRIPTION
**Pull request by inFocus7**
**What this PR does / why we need it**:
We currently run Trivy scans during [test workflow runs](https://github.com/eraser-dev/eraser/blob/3febb3a2f1345812d7b17071c700fc30ceaf398e/.github/workflows/test.yaml#L137C1-L186), but want to run them more frequently. 

This adds a cron job to run Trivy scans every Monday at 7:00 AM UTC. Those scans are ran against eraser images built on `main` branch.

It also allows for manual dispatching to do above _and_ allows running scans of specific release versions.

Each run's vulnerability can also be published to our security tab (I couldn't find a way to make them into issues) if either ran as part of the cron/schedule, or if in our manually dispatch we enable it.

I also added some workflow documentation, only information regarding this action is there, so any new developers can understand it better (hopefully).

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #783 

**Special notes for your reviewer**:
- The matrix is a teeny bit weird. Mainly because each build command is different and the environment variable needed for them is as well so I added those necessary variables to the matrix.
- Ran the scans using the current code in a [separate fork](https://github.com/not-fabian/eraser/actions/workflows/scan-images.yaml)
- Results (that had vulnerabilities) are recorded in the [security tab](https://github.com/not-fabian/eraser/security/code-scanning?query=is%3Aopen+branch%3Amain+sort%3Acreated-desc)
    - If you see a lot added by **Trivy** in my results, it is likely scanned from v1.1.0 image, since I wanted our oldest release with the highest possible CVEs.

_example of a vulnerability caught and recorded, scanning our released v1.1.0 images_
_as far as i've seen we can't really update the title or metadata, as these are generated straight from the SERIF file generated by the trivy results_
Screen Shot 2023-09-07 at 7 44 09 PM

There are three runs:
- One leaving the `version` input blank (default), which should be building aginst the branch it ran against, then testing those images.
- One setting the `version` to the latest beta release, which skips building images, and scans those images.
- A scheduled run

### CODE REVIEW

File: .github/workflows/scan-images.yaml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        with:
+          egress-policy: audit
+      - name: Download trivy
+        run: |
+          pushd $(mktemp -d)
+          wget https://github.com/aquasecurity/trivy/releases/download/v${{ env.TRIVY_VERSION }}/trivy_${{ env.TRIVY_VERSION }}_Linux-64bit.tar.gz
+          tar zxvf trivy_${{ env.TRIVY_VERSION }}_Linux-64bit.tar.gz
+          echo "$(pwd)" >> $GITHUB_PATH
+        env:
+          TRIVY_VERSION: "0.28.0"
```
Comment (sozercan) [inline .github/workflows/scan-images.yaml]:
```suggestion
          TRIVY_VERSION: "0.45.0"
```

File: .github/workflows/scan-images.yaml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    if: github.event_name == 'schedule' || (github.event_name == 'workflow_dispatch' && github.event.inputs.scan_main == true)
+    runs-on: ubuntu-latest
+    needs: setup_trivy
+    timeout-minutes: 15
+    steps:
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@f43a0e5ff2bd294095638e18286ca9a3d1956744 # v3.6.0
+      - name: Get repo
+        run: |
+          echo "REPO=$(echo $GITHUB_REPOSITORY | awk '{print tolower($0)}')" >> $GITHUB_ENV
```
Comment (sozercan) [inline .github/workflows/scan-images.yaml]: do we need this? previously org name was capitalized which docker didn't support

Comment (not-fabian) [reply .github/workflows/scan-images.yaml]: Good call, not necessary 👍 Removed using the `$GITHUB_REPOSITORY` overall, but can re-add if wanted.

File: .github/workflows/scan-images.yaml (hunk 1 above)
Comment (not-fabian) [reply .github/workflows/scan-images.yaml]: Switched to using the [trivy github action](https://github.com/aquasecurity/trivy-action) for our scans. Pinned the latest released version which [uses trivy 0.45.0](https://github.com/aquasecurity/trivy-action/commit/fbd16365eb88e12433951383f5e99bd901fc618f).

File: .github/workflows/scan-images.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      - name: Scan remover for vulnerabilities
+        uses: aquasecurity/trivy-action@0.12.0
+        with:
+          image-ref: ${{ env.REGISTRY }}/${{ env.REGISTRY_POSTFIX }}/remover:${{ env.TAG }}
+          vuln-type: 'os,library'
+          format: 'table'
+          ignore-unfixed: true
+          exit-code: 1
+      - name: Scan eraser-manager for vulnerabilities
+        uses: aquasecurity/trivy-action@0.12.0
```
Comment (ashnamehrotra) [inline .github/workflows/scan-images.yaml]: pin trivy action (trivy-action@fbd16365eb88e12433951383f5e99bd901fc618f #0.12.0)

Comment (ashnamehrotra) [reply .github/workflows/scan-images.yaml]: same for others

File: .github/workflows/scan-images.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    name: Scan for vulnerabilities ${{ github.event_name == 'schedule' && '(scheduled)' || '' }}
+    runs-on: ubuntu-latest
+    env:
+      # Scanning RELEASED versions require the project `eraser-dev` as part of the repo name.
+      REGISTRY_POSTFIX: ${{ github.event.inputs.version == '' && 'eraser-test' || 'eraser-dev' }}
+      TAG: ${{ github.event.inputs.version == '' && 'test' || github.event.inputs.version }}
+    timeout-minutes: 15
+    steps:
+      - name: Harden Runner
+        uses: step-security/harden-runner@8ca2b8b2ece13480cda6dacd3511b49857a23c09
```
Comment (ashnamehrotra) [inline .github/workflows/scan-images.yaml]: add # v2.5.1

File: .github/workflows/scan-images.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          ignore-unfixed: true
+          exit-code: 1
+      - name: Scan trivy-scanner for vulnerabilities
+        uses: aquasecurity/trivy-action@0.12.0
+        with:
+          image-ref: ${{ env.REGISTRY }}/${{ env.REGISTRY_POSTFIX }}/eraser-trivy-scanner:${{ env.TAG }}
+          vuln-type: 'os,library'
+          format: 'table'
+          ignore-unfixed: true
+          exit-code: 1
```
Comment (ashnamehrotra) [inline .github/workflows/scan-images.yaml]: if there are vulnerabilities, would it be possible to have the workflow create an issue to update?

Comment (inFocus7) [reply .github/workflows/scan-images.yaml]: I couldn't find any examples of creating an issue of the scan results.

I did see a [section of the trivy-action docs](https://github.com/aquasecurity/trivy-action#using-trivy-with-github-code-scanning) saying it's possible to upload the results to **[Github Security](https://github.com/eraser-dev/eraser/security)** using `github/codeql-action`. Not sure if you think we {s/c}hould follow that path 👀. If so, I think I can add it to my other fork and test it out. 

An issue I just thought of is: "Will this workflow  instantly exit on the step where the **first** instance of a Trivy scan fails?". I'm going to see what happens by using my other fork to scan some vulnerable images, to make sure it scans every image even if all are vulnerable 👍🏼

Comment (inFocus7) [reply .github/workflows/scan-images.yaml]: The easiest/documented way I found of doing this was to use codeql to upload to GitHub Security tab. I made the updates to do so, although if we want to create an action - or postpone creating an action - and remove this, I can do so 👍🏼 

In the [security tab](https://github.com/not-fabian/eraser/security/code-scanning?query=is%3Aopen+branch%3Amain+sort%3Acreated-desc) we can see all the vulns caught for a scan I did on 1.1.0. Did it on the first release I saw to catch as many as I could.

It's a teeny bit confusing going through them although I didnt' find a way to modify their name, I think it's based on the Trivy-generated SERIF. Although going into one of the vulns caught, we can see that the metadata included the run it was part of: https://github.com/not-fabian/eraser/security/code-scanning/37

**Comment by inFocus7** [comment]
The trivy image we use ([44.0](https://github.com/inFocus7/eraser/blob/61df001095b7a942ee0f4d6dd3b6a66632014ce9/Dockerfile#L4)) has a [vulnerability](https://github.com/eraser-dev/eraser/actions/runs/6116021398/job/16600577720?pr=858). I ran scans against 44.1 & 45.0 (latest release) and both of those have it as well, so no version we can upgrade to (yet).

Not sure if we want to skip the trivy scan in the test for this pr (unsure if possible). I submitted a report just in case it hasn't been noticed yet on Trivy's end. If it gets fixed tomorrow that'll be very :nice:

---

**Update**
Was [bumped with chore bot](https://github.com/aquasecurity/trivy/commit/48cbf45534155b6be1dc467a4e30564acb2bb2e4), so we'll just need to wait on a Trivy release, then bump our build of it. If we want to ignore it. If so, we may be able to create a `.trivyignore` file in root and add the vulnerability, but probably best to wait and not to have any.

Also not sure if backports are done in Trivy or how often, so worst-case we'd need to do a major bump from 44.X to 45.X.

File: .github/workflows/scan-images.yaml (hunk 5 above)
Comment (ashnamehrotra) [reply .github/workflows/scan-images.yaml]: I think the security tab link is private, but from the screenshot it looks good!

**Comment by ashnamehrotra** [comment]
> The trivy image we use ([44.0](https://github.com/inFocus7/eraser/blob/61df001095b7a942ee0f4d6dd3b6a66632014ce9/Dockerfile#L4)) has a [vulnerability](https://github.com/eraser-dev/eraser/actions/runs/6116021398/job/16600577720?pr=858). I ran scans against 44.1 & 45.0 (latest release) and both of those have it as well, so no version we can upgrade to (yet).
> 
> Not sure if we want to skip the trivy scan in the test for this pr (unsure if possible). I submitted a report just in case it hasn't been noticed yet on Trivy's end. If it gets fixed tomorrow that'll be very :nice:
> 
> **Update** Was [bumped with chore bot](https://github.com/aquasecurity/trivy/commit/48cbf45534155b6be1dc467a4e30564acb2bb2e4), so we'll just need to wait on a Trivy release, then bump our build of it. If we want to ignore it. If so, we may be able to create a `.trivyignore` file in root and add the vulnerability, but probably best to wait and not to have any.
> 
> Also not sure if backports are done in Trivy or how often, so worst-case we'd need to do a major bump from 44.X to 45.X.

We can wait for the trivy release to fix this, thanks for checking!

**Comment by inFocus7** [comment]
@ashnamehrotra 

Trivy just released a new version (0.45.1). Scanned it locally and no CVEs, so it should be good to pass now! 🎆

**Comment by ashnamehrotra** [comment]
@inFocus7 you will also need to update the go.mod for the test to pass

**Comment by inFocus7** [comment]
@ashnamehrotra Thanks! I'll update if it fails after this run where I saw I didn't update the tag in Makefile. 
Doing a bump of trivy to `0.45.1` in go.mod led to a lot of k8s updates. I see we have `replace`s in hand, but didn't feel comfortable/want to risk break anything yet, if not necessary 😅 

## Update
Looks like the Trivy scan failed again. It is no longer due to the `cyphar/filepath-securejoin` CVE, which is good since it means it's using the latest Trivy release (0.45.1)... **But**, the updated Trivy image (which I also scanned locally `trivy image ghcr.io/aquasecurity/trivy:0.45.1`) has a CVE. 😢 

Trying to think of workarounds:
- Add a `.trivyignore` to ignore the CVE.
  - Good for now, but we'd want to update again it in the future once Trivy updates and CVE is good.
- Skip or remove the Trivy scan portion of the test.
- Don't skip, but don't fail/block CI on check failure. 
  - I feel like it _should_ be possible to make the check optional, but not 100% certain at this moment.

**Comment by ashnamehrotra** [comment]
@inFocus7 I added the trivy ignore for now, thank you!

**Review by ashnamehrotra — APPROVED** [review]

**Pull request merged**